{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting category_encoders==2.5.0\n",
      "  Downloading category_encoders-2.5.0-py2.py3-none-any.whl (69 kB)\n",
      "\u001b[K     |████████████████████████████████| 69 kB 10.3 MB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: pandas>=1.0.5 in /home/labuser/.local/lib/python3.8/site-packages (from category_encoders==2.5.0) (1.3.0)\n",
      "Requirement already satisfied: statsmodels>=0.9.0 in /home/labuser/.local/lib/python3.8/site-packages (from category_encoders==2.5.0) (0.12.2)\n",
      "Requirement already satisfied: scikit-learn>=0.20.0 in /home/labuser/.local/lib/python3.8/site-packages (from category_encoders==2.5.0) (1.0.2)\n",
      "Requirement already satisfied: scipy>=1.0.0 in /home/labuser/.local/lib/python3.8/site-packages (from category_encoders==2.5.0) (1.7.0)\n",
      "Requirement already satisfied: patsy>=0.5.1 in /home/labuser/.local/lib/python3.8/site-packages (from category_encoders==2.5.0) (0.5.2)\n",
      "Requirement already satisfied: numpy>=1.14.0 in /home/labuser/.local/lib/python3.8/site-packages (from category_encoders==2.5.0) (1.21.1)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in /home/labuser/.local/lib/python3.8/site-packages (from pandas>=1.0.5->category_encoders==2.5.0) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2017.3 in /usr/lib/python3/dist-packages (from pandas>=1.0.5->category_encoders==2.5.0) (2019.3)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /home/labuser/.local/lib/python3.8/site-packages (from scikit-learn>=0.20.0->category_encoders==2.5.0) (3.1.0)\n",
      "Requirement already satisfied: joblib>=0.11 in /home/labuser/.local/lib/python3.8/site-packages (from scikit-learn>=0.20.0->category_encoders==2.5.0) (1.1.0)\n",
      "Requirement already satisfied: six in /usr/lib/python3/dist-packages (from patsy>=0.5.1->category_encoders==2.5.0) (1.14.0)\n",
      "Installing collected packages: category-encoders\n",
      "  Attempting uninstall: category-encoders\n",
      "    Found existing installation: category-encoders 2.4.0\n",
      "    Uninstalling category-encoders-2.4.0:\n",
      "      Successfully uninstalled category-encoders-2.4.0\n",
      "Successfully installed category-encoders-2.5.0\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Collecting hyperopt==0.2.7\n",
      "  Downloading hyperopt-0.2.7-py2.py3-none-any.whl (1.6 MB)\n",
      "\u001b[K     |████████████████████████████████| 1.6 MB 32.8 MB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: six in /usr/lib/python3/dist-packages (from hyperopt==0.2.7) (1.14.0)\n",
      "Requirement already satisfied: networkx>=2.2 in /home/labuser/.local/lib/python3.8/site-packages (from hyperopt==0.2.7) (2.3)\n",
      "Collecting py4j\n",
      "  Downloading py4j-0.10.9.7-py2.py3-none-any.whl (200 kB)\n",
      "\u001b[K     |████████████████████████████████| 200 kB 108.6 MB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: scipy in /home/labuser/.local/lib/python3.8/site-packages (from hyperopt==0.2.7) (1.7.0)\n",
      "Requirement already satisfied: cloudpickle in /home/labuser/.local/lib/python3.8/site-packages (from hyperopt==0.2.7) (2.1.0)\n",
      "Requirement already satisfied: tqdm in /home/labuser/.local/lib/python3.8/site-packages (from hyperopt==0.2.7) (4.64.0)\n",
      "Requirement already satisfied: numpy in /home/labuser/.local/lib/python3.8/site-packages (from hyperopt==0.2.7) (1.21.1)\n",
      "Requirement already satisfied: future in /home/labuser/.local/lib/python3.8/site-packages (from hyperopt==0.2.7) (0.18.2)\n",
      "Requirement already satisfied: decorator>=4.3.0 in /home/labuser/.local/lib/python3.8/site-packages (from networkx>=2.2->hyperopt==0.2.7) (5.1.1)\n",
      "Installing collected packages: py4j, hyperopt\n",
      "\u001b[33m  WARNING: The script hyperopt-mongo-worker is installed in '/home/labuser/.local/bin' which is not on PATH.\n",
      "  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.\u001b[0m\n",
      "Successfully installed hyperopt-0.2.7 py4j-0.10.9.7\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Requirement already satisfied: lightgbm==3.3.0 in /home/labuser/.local/lib/python3.8/site-packages (3.3.0)\n",
      "Requirement already satisfied: scipy in /home/labuser/.local/lib/python3.8/site-packages (from lightgbm==3.3.0) (1.7.0)\n",
      "Requirement already satisfied: wheel in /usr/lib/python3/dist-packages (from lightgbm==3.3.0) (0.34.2)\n",
      "Requirement already satisfied: numpy in /home/labuser/.local/lib/python3.8/site-packages (from lightgbm==3.3.0) (1.21.1)\n",
      "Requirement already satisfied: scikit-learn!=0.22.0 in /home/labuser/.local/lib/python3.8/site-packages (from lightgbm==3.3.0) (1.0.2)\n",
      "Requirement already satisfied: joblib>=0.11 in /home/labuser/.local/lib/python3.8/site-packages (from scikit-learn!=0.22.0->lightgbm==3.3.0) (1.1.0)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /home/labuser/.local/lib/python3.8/site-packages (from scikit-learn!=0.22.0->lightgbm==3.3.0) (3.1.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Collecting matplotlib==3.5.1\n",
      "  Downloading matplotlib-3.5.1-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.whl (11.3 MB)\n",
      "\u001b[K     |████████████████████████████████| 11.3 MB 36.7 MB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: packaging>=20.0 in /home/labuser/.local/lib/python3.8/site-packages (from matplotlib==3.5.1) (21.3)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /home/labuser/.local/lib/python3.8/site-packages (from matplotlib==3.5.1) (9.0.0)\n",
      "Requirement already satisfied: pyparsing>=2.2.1 in /home/labuser/.local/lib/python3.8/site-packages (from matplotlib==3.5.1) (3.0.9)\n",
      "Requirement already satisfied: numpy>=1.17 in /home/labuser/.local/lib/python3.8/site-packages (from matplotlib==3.5.1) (1.21.1)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /home/labuser/.local/lib/python3.8/site-packages (from matplotlib==3.5.1) (1.4.2)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/labuser/.local/lib/python3.8/site-packages (from matplotlib==3.5.1) (0.11.0)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /home/labuser/.local/lib/python3.8/site-packages (from matplotlib==3.5.1) (2.8.2)\n",
      "Collecting fonttools>=4.22.0\n",
      "  Downloading fonttools-4.39.2-py3-none-any.whl (1.0 MB)\n",
      "\u001b[K     |████████████████████████████████| 1.0 MB 91.9 MB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: six>=1.5 in /usr/lib/python3/dist-packages (from python-dateutil>=2.7->matplotlib==3.5.1) (1.14.0)\n",
      "Installing collected packages: fonttools, matplotlib\n",
      "\u001b[33m  WARNING: The scripts fonttools, pyftmerge, pyftsubset and ttx are installed in '/home/labuser/.local/bin' which is not on PATH.\n",
      "  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.\u001b[0m\n",
      "  Attempting uninstall: matplotlib\n",
      "    Found existing installation: matplotlib 3.4.3\n",
      "    Uninstalling matplotlib-3.4.3:\n",
      "      Successfully uninstalled matplotlib-3.4.3\n",
      "Successfully installed fonttools-4.39.2 matplotlib-3.5.1\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Collecting numpy==1.22.4\n",
      "  Downloading numpy-1.22.4-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (16.9 MB)\n",
      "\u001b[K     |████████████████████████████████| 16.9 MB 21.7 MB/s eta 0:00:01\n",
      "\u001b[31mERROR: visions 0.7.4 has requirement networkx>=2.4, but you'll have networkx 2.3 which is incompatible.\u001b[0m\n",
      "\u001b[31mERROR: numba 0.55.1 has requirement numpy<1.22,>=1.18, but you'll have numpy 1.22.4 which is incompatible.\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: numpy\n",
      "  Attempting uninstall: numpy\n",
      "    Found existing installation: numpy 1.21.1\n",
      "    Uninstalling numpy-1.21.1:\n",
      "      Successfully uninstalled numpy-1.21.1\n",
      "\u001b[33m  WARNING: The scripts f2py, f2py3 and f2py3.8 are installed in '/home/labuser/.local/bin' which is not on PATH.\n",
      "  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.\u001b[0m\n",
      "Successfully installed numpy-1.22.4\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Requirement already satisfied: pandas==1.3.0 in /home/labuser/.local/lib/python3.8/site-packages (1.3.0)\n",
      "Requirement already satisfied: pytz>=2017.3 in /usr/lib/python3/dist-packages (from pandas==1.3.0) (2019.3)\n",
      "Requirement already satisfied: numpy>=1.17.3 in /home/labuser/.local/lib/python3.8/site-packages (from pandas==1.3.0) (1.22.4)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in /home/labuser/.local/lib/python3.8/site-packages (from pandas==1.3.0) (2.8.2)\n",
      "Requirement already satisfied: six>=1.5 in /usr/lib/python3/dist-packages (from python-dateutil>=2.7.3->pandas==1.3.0) (1.14.0)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Collecting scikit_learn==1.1.2\n",
      "  Downloading scikit_learn-1.1.2-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (31.2 MB)\n",
      "\u001b[K     |████████████████████████████████| 31.2 MB 74 kB/s s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: joblib>=1.0.0 in /home/labuser/.local/lib/python3.8/site-packages (from scikit_learn==1.1.2) (1.1.0)\n",
      "Requirement already satisfied: numpy>=1.17.3 in /home/labuser/.local/lib/python3.8/site-packages (from scikit_learn==1.1.2) (1.22.4)\n",
      "Requirement already satisfied: scipy>=1.3.2 in /home/labuser/.local/lib/python3.8/site-packages (from scikit_learn==1.1.2) (1.7.0)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /home/labuser/.local/lib/python3.8/site-packages (from scikit_learn==1.1.2) (3.1.0)\n",
      "Installing collected packages: scikit-learn\n",
      "  Attempting uninstall: scikit-learn\n",
      "    Found existing installation: scikit-learn 1.0.2\n",
      "    Uninstalling scikit-learn-1.0.2:\n",
      "      Successfully uninstalled scikit-learn-1.0.2\n",
      "Successfully installed scikit-learn-1.1.2\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Collecting seaborn==0.12.0\n",
      "  Downloading seaborn-0.12.0-py3-none-any.whl (285 kB)\n",
      "\u001b[K     |████████████████████████████████| 285 kB 30.2 MB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: pandas>=0.25 in /home/labuser/.local/lib/python3.8/site-packages (from seaborn==0.12.0) (1.3.0)\n",
      "Requirement already satisfied: numpy>=1.17 in /home/labuser/.local/lib/python3.8/site-packages (from seaborn==0.12.0) (1.22.4)\n",
      "Requirement already satisfied: matplotlib>=3.1 in /home/labuser/.local/lib/python3.8/site-packages (from seaborn==0.12.0) (3.5.1)\n",
      "Requirement already satisfied: pytz>=2017.3 in /usr/lib/python3/dist-packages (from pandas>=0.25->seaborn==0.12.0) (2019.3)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in /home/labuser/.local/lib/python3.8/site-packages (from pandas>=0.25->seaborn==0.12.0) (2.8.2)\n",
      "Requirement already satisfied: pyparsing>=2.2.1 in /home/labuser/.local/lib/python3.8/site-packages (from matplotlib>=3.1->seaborn==0.12.0) (3.0.9)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /home/labuser/.local/lib/python3.8/site-packages (from matplotlib>=3.1->seaborn==0.12.0) (4.39.2)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /home/labuser/.local/lib/python3.8/site-packages (from matplotlib>=3.1->seaborn==0.12.0) (9.0.0)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/labuser/.local/lib/python3.8/site-packages (from matplotlib>=3.1->seaborn==0.12.0) (21.3)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/labuser/.local/lib/python3.8/site-packages (from matplotlib>=3.1->seaborn==0.12.0) (0.11.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /home/labuser/.local/lib/python3.8/site-packages (from matplotlib>=3.1->seaborn==0.12.0) (1.4.2)\n",
      "Requirement already satisfied: six>=1.5 in /usr/lib/python3/dist-packages (from python-dateutil>=2.7.3->pandas>=0.25->seaborn==0.12.0) (1.14.0)\n",
      "Installing collected packages: seaborn\n",
      "  Attempting uninstall: seaborn\n",
      "    Found existing installation: seaborn 0.11.2\n",
      "    Uninstalling seaborn-0.11.2:\n",
      "      Successfully uninstalled seaborn-0.11.2\n",
      "Successfully installed seaborn-0.12.0\n",
      "Note: you may need to restart the kernel to use updated packages.\n",
      "Requirement already satisfied: shap==0.40.0 in /home/labuser/.local/lib/python3.8/site-packages (0.40.0)\n",
      "Requirement already satisfied: pandas in /home/labuser/.local/lib/python3.8/site-packages (from shap==0.40.0) (1.3.0)\n",
      "Requirement already satisfied: scikit-learn in /home/labuser/.local/lib/python3.8/site-packages (from shap==0.40.0) (1.1.2)\n",
      "Requirement already satisfied: slicer==0.0.7 in /home/labuser/.local/lib/python3.8/site-packages (from shap==0.40.0) (0.0.7)\n",
      "Requirement already satisfied: numpy in /home/labuser/.local/lib/python3.8/site-packages (from shap==0.40.0) (1.22.4)\n",
      "Requirement already satisfied: cloudpickle in /home/labuser/.local/lib/python3.8/site-packages (from shap==0.40.0) (2.1.0)\n",
      "Requirement already satisfied: tqdm>4.25.0 in /home/labuser/.local/lib/python3.8/site-packages (from shap==0.40.0) (4.64.0)\n",
      "Requirement already satisfied: scipy in /home/labuser/.local/lib/python3.8/site-packages (from shap==0.40.0) (1.7.0)\n",
      "Requirement already satisfied: numba in /home/labuser/.local/lib/python3.8/site-packages (from shap==0.40.0) (0.55.1)\n",
      "Requirement already satisfied: packaging>20.9 in /home/labuser/.local/lib/python3.8/site-packages (from shap==0.40.0) (21.3)\n",
      "Requirement already satisfied: pytz>=2017.3 in /usr/lib/python3/dist-packages (from pandas->shap==0.40.0) (2019.3)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in /home/labuser/.local/lib/python3.8/site-packages (from pandas->shap==0.40.0) (2.8.2)\n",
      "Requirement already satisfied: joblib>=1.0.0 in /home/labuser/.local/lib/python3.8/site-packages (from scikit-learn->shap==0.40.0) (1.1.0)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /home/labuser/.local/lib/python3.8/site-packages (from scikit-learn->shap==0.40.0) (3.1.0)\n",
      "Requirement already satisfied: setuptools in /usr/lib/python3/dist-packages (from numba->shap==0.40.0) (45.2.0)\n",
      "Requirement already satisfied: llvmlite<0.39,>=0.38.0rc1 in /home/labuser/.local/lib/python3.8/site-packages (from numba->shap==0.40.0) (0.38.1)\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /home/labuser/.local/lib/python3.8/site-packages (from packaging>20.9->shap==0.40.0) (3.0.9)\n",
      "Requirement already satisfied: six>=1.5 in /usr/lib/python3/dist-packages (from python-dateutil>=2.7.3->pandas->shap==0.40.0) (1.14.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "# install libraries\n",
    "%pip install category_encoders==2.5.0\n",
    "%pip install hyperopt==0.2.7\n",
    "%pip install lightgbm==3.3.0\n",
    "%pip install matplotlib==3.5.1\n",
    "%pip install numpy==1.22.4\n",
    "%pip install pandas==1.3.0\n",
    "%pip install scikit_learn==1.1.2\n",
    "%pip install seaborn==0.12.0\n",
    "%pip install shap==0.40.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting numba==0.53.1\n",
      "  Downloading numba-0.53.1-cp38-cp38-manylinux2014_x86_64.whl (3.4 MB)\n",
      "\u001b[K     |████████████████████████████████| 3.4 MB 29.8 MB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: setuptools in /usr/lib/python3/dist-packages (from numba==0.53.1) (45.2.0)\n",
      "Requirement already satisfied: numpy>=1.15 in /home/labuser/.local/lib/python3.8/site-packages (from numba==0.53.1) (1.22.4)\n",
      "Collecting llvmlite<0.37,>=0.36.0rc1\n",
      "  Downloading llvmlite-0.36.0-cp38-cp38-manylinux2010_x86_64.whl (25.3 MB)\n",
      "\u001b[K     |████████████████████████████████| 25.3 MB 92.6 MB/s eta 0:00:01\n",
      "\u001b[?25hInstalling collected packages: llvmlite, numba\n",
      "  Attempting uninstall: llvmlite\n",
      "    Found existing installation: llvmlite 0.38.1\n",
      "    Uninstalling llvmlite-0.38.1:\n",
      "      Successfully uninstalled llvmlite-0.38.1\n",
      "  Attempting uninstall: numba\n",
      "    Found existing installation: numba 0.55.1\n",
      "    Uninstalling numba-0.55.1:\n",
      "      Successfully uninstalled numba-0.55.1\n",
      "Successfully installed llvmlite-0.36.0 numba-0.53.1\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install numba==0.53.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import os\n",
    "path = os.getcwd()\n",
    "import sys\n",
    "sys.path.append(\"\")\n",
    "from utils import *\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import random\n",
    "\n",
    "import category_encoders as ce\n",
    "import lightgbm as lgb\n",
    "from lightgbm import LGBMClassifier\n",
    "\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.metrics import roc_curve\n",
    "from sklearn.metrics import auc\n",
    "from sklearn.metrics import precision_recall_curve\n",
    "from sklearn.metrics import precision_score\n",
    "from sklearn.metrics import average_precision_score\n",
    "import bisect\n",
    "\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "\n",
    "from hyperopt import fmin, tpe, hp, anneal, Trials\n",
    "\n",
    "import os\n",
    "import gc\n",
    "\n",
    "import datetime\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "pd.set_option('display.float_format', lambda x: '%.3f' % x)\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
